File: apps/lesson-plan-generator/internal/pipeline/pipeline.go
// Package pipeline orchestrates the lesson-plan generation steps.
package pipeline

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"lesson-plan-generator/internal/domain"
)

// Generator holds configuration and runs the full plan pipeline.
type Generator struct {
	apiKey     string
	voice      string
	outputDir  string
	httpClient *http.Client
}

func New(apiKey, voice, outputDir string) *Generator {
	return &Generator{
		apiKey:     apiKey,
		voice:      voice,
		outputDir:  outputDir,
		httpClient: &http.Client{Timeout: 60 * time.Second},
	}
}

func (g *Generator) GeneratePlan(ctx context.Context, lessonID string) (domain.VideoPlan, error) {
	// 1. Load lesson
	lesson, err := loadLesson(lessonID)
	if err != nil {
		return domain.VideoPlan{}, fmt.Errorf("load: %w", err)
	}

	// 2. Split narration
	segments := splitNarration(lesson)

	// 3 & 4. Generate TTS + measure durations
	audioSegments, err := g.generateTTS(ctx, segments, lessonID)
	if err != nil {
		return domain.VideoPlan{}, fmt.Errorf("tts: %w", err)
	}

	// Concatenate segments into full audio track
	fullAudioFile, err := concatenateAudio(audioSegments, lessonID, g.outputDir)
	if err != nil {
		return domain.VideoPlan{}, fmt.Errorf("concatenate: %w", err)
	}

	// 5. Map scene scripts → (script, duration)
	sceneData := generateScenes(lesson.SceneScripts, audioSegments)

	// 6. Compute video timeline + assign audio start times
	scenes, timedSegments, err := computeTimeline(sceneData, audioSegments)
	if err != nil {
		return domain.VideoPlan{}, fmt.Errorf("timeline: %w", err)
	}

	// 7. Assemble final VideoPlan
	return assemble(lessonID, timedSegments, fullAudioFile, scenes), nil
}

File: apps/lesson-plan-generator/app/pipeline.py
"""Orchestrates all pipeline steps in order."""

from pathlib import Path

from openai import AsyncOpenAI

from app.models import VideoPlan
from app.steps.assembler import assemble
from app.steps.loader import load_lesson
from app.steps.narration import split_narration
from app.steps.scenes import generate_scenes
from app.steps.timeline import compute_timeline
from app.steps.tts import concatenate_audio, generate_tts


async def generate_plan(
    lesson_id: str,
    output_dir: Path,
    openai_client: AsyncOpenAI,
    tts_voice: str = "alloy",
) -> VideoPlan:
    # 1. Load lesson
    lesson = load_lesson(lesson_id)

    # 2. Split narration into segments
    narration_segments = split_narration(lesson)

    # 3 & 4. Generate TTS audio files + measure durations
    audio_segments = await generate_tts(
        narration_segments, lesson_id, output_dir, openai_client, tts_voice
    )

    # Concatenate segments into full audio track
    full_audio_file = concatenate_audio(audio_segments, lesson_id, output_dir)

    # 5. Map scene scripts → (script, duration)
    scene_data = generate_scenes(lesson.scene_scripts, audio_segments)

    # 6. Compute video timeline + assign audio start times
    scenes, timed_segments = compute_timeline(scene_data, audio_segments)

    # 7. Assemble final VideoPlan
    return assemble(lesson_id, timed_segments, full_audio_file, scenes)
